src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+        if float(norm) > precision_tol:
+            return False
+        elif len(self.partition) == 2:
+            return True
+        else:
+            # Unital is defined for quantum channels only. But we can extend it to quantum combs as follows:
+            return QuantumChannel(  # pragma: no cover
+                sub_comb, self.partition[2:], pure=False, backend=self._backend
+            ).is_unital(order, precision_tol)
+
+    def is_channel(
+        self,
+        order: Optional[Union[int, str]] = None,
+        precision_tol_causal: float = 1e-8,
+        precision_tol_psd: float = 1e-8,
+    ):
+        """Returns bool indicating if Choi operator :math:`\\mathcal{E}` is a channel.
+
+        Args:
+            order (int or str, optional): order of the norm used to calculate causality.
+                Defaults to ``None``.
+            precision_tol_causal (float, optional): threshold :math:`\\epsilon` that defines if
+                Choi operator of the network is :math:`\\epsilon`-close to causality in the norm
+                given by ``order``. Defaults to :math:`10^{-8}`.
+            precision_tol_psd (float, optional): threshold value used to check if eigenvalues of
+                the Choi operator :math:`\\mathcal{E}` are such that
+                :math:`\\textup{eigenvalues}(\\mathcal{E}) >= \\textup{precision_tol_psd}`.
+                Note that this parameter can be set to negative values.
+                Defaults to :math:`0.0`.
+
+        Returns:
+            bool: Channel condition.
+        """
+        return self.is_causal(
+            order, precision_tol_causal
+        ) and self.is_positive_semidefinite(precision_tol_psd)
+
+    def apply(self, state):
+        """Apply the Choi operator :math:`\\mathcal{E}` to ``state`` :math:`\\varrho`.
+
+        It is assumed that ``state`` :math:`\\varrho` is a density matrix.
+
+        Args:
+            state (ndarray): density matrix of a ``state``.
+
+        Returns:
+            ndarray: Resulting state :math:`\\mathcal{E}(\\varrho)`.
+        """
+        operator = self.copy().operator()
+        conj = self._backend.np.conj
 
         if self.is_pure():
-            matrix = self._einsum("jk,lm -> kjml", matrix, np.conj(matrix))
+            return self._einsum("ij,lk,il", operator, conj(operator), state)
+
+        return self._einsum("ijkl, jl", operator, state)
+
+
+class StochQuantumNetwork:
+    pass
 
-        return matrix
 
-    def _check_subscript_pattern(self, subscripts: str):
-        """Checks if input subscript match any implemented pattern."""
-        braket = "[a-z]"
-        pattern_two = re.compile(braket * 2 + "," + braket * 2 + "->" + braket * 2)
-        pattern_four = re.compile(braket * 4 + "," + braket * 2 + "->" + braket * 2)
+def link_product(
+    subscripts: str = "ij,jk -> ik",
+    *operands: QuantumNetwork,
+    backend=None,
+    surpress_warning=False,
+    casting: bool = True,
```

Comment:
It seems `casting` is not needed.

Reply:
Ah, sure. I tried to implement the functions that cast the output of the link product to quantum channels if possible. But later I found the casting may mislead the users and I need to check many complex cases to prevent bugs. This functionality is removed now.